File: odoo_surface_mcp/tools/discovery.py
"""Layer 1 — Discovery tools: get_models, get_model_actions."""
import xml.etree.ElementTree as ET
from typing import Optional

from mcp.server.fastmcp import FastMCP

from odoo_surface_mcp.cache import Cache
from odoo_surface_mcp.odoo_client import OdooClient

_BUTTON_TYPES = {"object", "action"}
_RELATIONAL_TYPES = {"many2one", "many2many", "one2many"}


def register(mcp: FastMCP, client: OdooClient, cache: Cache) -> None:

    @mcp.tool(description=(
        "List primary models the user can navigate to via menus "
        "(get_models()), or list relational models reachable from a base "
        "model via its form-view fields (get_models(base='sale.order'))."
    ))
    def get_models(base: Optional[str] = None) -> list[dict]:
        if base:
            return _related_models(client, cache, base)
        return _primary_models(client, cache)

    @mcp.tool(description=(
        "Return all actions available on a model: server actions (Action menu), "
        "report actions (Print menu), and form-view buttons (type=object/action). "
        "Also returns CRUD access flags for the current user."
    ))
    def get_model_actions(model: str, action_id: Optional[int] = None) -> dict:
        cache_key = f"model_actions:{model}:{action_id}"
        cached = cache.get(cache_key)
        if cached:
            return cached

        result = _collect_model_actions(client, model)
        cache.set(cache_key, result)
        return result


# ---------------------------------------------------------------------------
# Internals
# ---------------------------------------------------------------------------

def _primary_models(client: OdooClient, cache: Cache) -> list[dict]:
    cached = cache.get("primary_models")
    if cached:
        return cached

    menus = client.execute(
        "ir.ui.menu", "search_read",
        [["action", "like", "ir.actions.act_window,"]],
        fields=["name", "complete_name", "action"],
    )

    act_ids = []
    menu_by_act: dict[int, dict] = {}
    for m in menus:
        action_ref = m.get("action") or ""
        if "," not in action_ref:
            continue
        act_id = int(action_ref.split(",", 1)[1])
        act_ids.append(act_id)
        menu_by_act.setdefault(act_id, m)  # keep first menu per action

    if not act_ids:
        return []

    actions = client.execute(
        "ir.actions.act_window", "search_read",
        [["id", "in", act_ids], ["res_model", "!=", False]],
        fields=["id", "name", "res_model", "view_mode"],
    )

    seen_models: set[str] = set()
    result = []
    for act in actions:
        model = act["res_model"]
        if model in seen_models:
            continue
        seen_models.add(model)
        menu = menu_by_act.get(act["id"], {})
        result.append({
            "model": model,
            "name": menu.get("name") or act["name"],
            "menu_path": menu.get("complete_name") or menu.get("name") or act["name"],
            "action_id": act["id"],
            "view_modes": act["view_mode"],
        })

    result.sort(key=lambda r: r["model"])
    cache.set("primary_models", result)
    return result


def _related_models(client: OdooClient, cache: Cache, base: str) -> list[dict]:
    cache_key = f"related_models:{base}"
    cached = cache.get(cache_key)
    if cached:
        return cached

    fields_meta = client.get_form_fields(base)
    result = []
    for fname, fmeta in fields_meta.items():
        if fmeta.get("type") not in _RELATIONAL_TYPES:
            continue
        relation = fmeta.get("relation")
        if not relation:
            continue
        result.append({
            "model": relation,
            "field": fname,
            "field_type": fmeta["type"],
            "label": fmeta.get("string", fname),
        })

    result.sort(key=lambda r: r["field"])
    cache.set(cache_key, result)
    return result


def _collect_model_actions(client: OdooClient, model: str) -> dict:
    """Collect all action sources for a model — mirrors the four places Odoo exposes actions.

    Odoo surfaces functional intents through four distinct mechanisms, each stored differently:

    1. Server actions (Action ▾ menu)
       Stored in ir.actions.server with binding_model_id and binding_type='action'.
       binding_view_types controls where they appear: 'form', 'list', or 'list,form'.

    2. Report actions (Print ▾ menu)
       Stored in ir.actions.report with binding_model_id.
       Same binding_view_types pattern as server actions.

    3. View buttons type="object" — the richest source of functional intents.
       NOT stored in ir.actions — they live in ir.ui.view.arch_db as XML attributes.
       They call Python methods directly (confirm, validate, cancel, publish, lock, ...).
       Visibility is controlled by the `invisible` attribute evaluated against the record.
       get_available_actions (Layer 2) evaluates these per-record using AST parsing.

    4. View buttons type="action" — less common.
       Also in arch_db. The button name is a numeric ir.actions.* id; calling it opens
       a related form or wizard rather than running a method directly.
    """
    # --- CRUD access ---
    access = {
        "can_create": client.check_access(model, "create"),
        "can_write":  client.check_access(model, "write"),
        "can_delete": client.check_access(model, "unlink"),
    }

    # --- ir.model id ---
    model_id = client.get_model_id(model)
    if not model_id:
        return {**access, "error": f"Model '{model}' not found in ir.model"}

    # --- Source 1: Server actions (Action ▾ menu) ---
    server_actions = client.execute(
        "ir.actions.server", "search_read",
        [["binding_model_id", "=", model_id], ["binding_type", "=", "action"]],
        fields=["id", "name", "binding_view_types", "state"],
    )

    # --- Source 2: Report actions (Print ▾ menu) ---
    reports = client.execute(
        "ir.actions.report", "search_read",
        [["binding_model_id", "=", model_id]],
        fields=["id", "name", "binding_view_types", "report_type"],
    )

    # --- Sources 3 & 4: View buttons (type=object / type=action) from form arch ---
    view_buttons = []
    try:
        arch = client.get_form_arch(model)
        root = ET.fromstring(arch)
        for btn in root.iter("button"):
            btn_type = btn.get("type", "")
            if btn_type not in _BUTTON_TYPES:
                continue
            view_buttons.append({
                "name": btn.get("name"),
                "label": btn.get("string") or btn.get("name"),
                "type": btn_type,
                "invisible": btn.get("invisible"),
                "groups": btn.get("groups"),
            })
    except Exception as exc:
        view_buttons = [{"error": str(exc)}]

    return {
        **access,
        "server_actions": [
            {"id": a["id"], "name": a["name"], "view_types": a["binding_view_types"]}
            for a in server_actions
        ],
        "reports": [
            {"id": r["id"], "name": r["name"], "view_types": r["binding_view_types"],
             "report_type": r["report_type"]}
            for r in reports
        ],
        "view_buttons": view_buttons,
    }

File: src/tools/discovery.ts
/** Layer 1 — Discovery tools: get_models, get_model_actions. */
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { OdooClient } from '../odooClient.js';
import { Cache } from '../cache.js';
import { xmlParser, FXPNode, iterNodes, ok } from '../utils.js';

const BUTTON_TYPES = new Set(['object', 'action']);
const RELATIONAL_TYPES = new Set(['many2one', 'many2many', 'one2many']);

// ─── Internals ────────────────────────────────────────────────────────────────

async function primaryModels(client: OdooClient, cache: Cache): Promise<unknown[]> {
  const cached = cache.get('primary_models');
  if (cached) return cached as unknown[];

  const menus = await client.execute('ir.ui.menu', 'search_read',
    [[['action', 'like', 'ir.actions.act_window,']]],
    { fields: ['name', 'complete_name', 'action'] },
  ) as Array<{ name: string; complete_name: string; action: string }>;

  const actIds: number[] = [];
  const menuByAct = new Map<number, typeof menus[0]>();
  for (const m of menus) {
    const ref = m.action ?? '';
    if (!ref.includes(',')) continue;
    const id = parseInt(ref.split(',')[1], 10);
    actIds.push(id);
    if (!menuByAct.has(id)) menuByAct.set(id, m);
  }
  if (!actIds.length) return [];

  const actions = await client.execute('ir.actions.act_window', 'search_read',
    [[['id', 'in', actIds], ['res_model', '!=', false]]],
    { fields: ['id', 'name', 'res_model', 'view_mode'] },
  ) as Array<{ id: number; name: string; res_model: string; view_mode: string }>;

  const seen = new Set<string>();
  const result: unknown[] = [];
  for (const act of actions) {
    if (seen.has(act.res_model)) continue;
    seen.add(act.res_model);
    const menu = menuByAct.get(act.id);
    result.push({
      model: act.res_model,
      name: menu?.name ?? act.name,
      menu_path: menu?.complete_name ?? menu?.name ?? act.name,
      action_id: act.id,
      view_modes: act.view_mode,
    });
  }
  (result as Array<{ model: string }>).sort((a, b) => a.model.localeCompare(b.model));
  cache.set('primary_models', result);
  return result;
}

async function relatedModels(client: OdooClient, cache: Cache, base: string): Promise<unknown[]> {
  const key = `related_models:${base}`;
  const cached = cache.get(key);
  if (cached) return cached as unknown[];

  const meta = await client.getFormFields(base) as Record<string, Record<string, unknown>>;
  const result: unknown[] = [];
  for (const [fname, fmeta] of Object.entries(meta)) {
    if (!RELATIONAL_TYPES.has(String(fmeta['type']))) continue;
    const relation = fmeta['relation'];
    if (!relation) continue;
    result.push({
      model: relation,
      field: fname,
      field_type: fmeta['type'],
      label: fmeta['string'] ?? fname,
    });
  }
  (result as Array<{ field: string }>).sort((a, b) => a.field.localeCompare(b.field));
  cache.set(key, result);
  return result;
}

export async function collectModelActions(client: OdooClient, model: string): Promise<Record<string, unknown>> {
  const access = {
    can_create: await client.checkAccess(model, 'create'),
    can_write: await client.checkAccess(model, 'write'),
    can_delete: await client.checkAccess(model, 'unlink'),
  };

  const modelId = await client.getModelId(model);
  if (!modelId) return { ...access, error: `Model '${model}' not found in ir.model` };

  const serverActions = await client.execute('ir.actions.server', 'search_read',
    [[['binding_model_id', '=', modelId], ['binding_type', '=', 'action']]],
    { fields: ['id', 'name', 'binding_view_types', 'state'] },
  ) as Array<{ id: number; name: string; binding_view_types: string }>;

  const reports = await client.execute('ir.actions.report', 'search_read',
    [[['binding_model_id', '=', modelId]]],
    { fields: ['id', 'name', 'binding_view_types', 'report_type'] },
  ) as Array<{ id: number; name: string; binding_view_types: string; report_type: string }>;

  let viewButtons: unknown[] = [];
  try {
    const arch = await client.getFormArch(model);
    const nodes = xmlParser.parse(arch) as FXPNode[];
    for (const node of iterNodes(nodes, 'button')) {
      const attrs = node[':@'] as Record<string, string> | undefined;
      if (!attrs) continue;
      const btnType = attrs['type'] ?? '';
      if (!BUTTON_TYPES.has(btnType)) continue;
      viewButtons.push({
        name: attrs['name'],
        label: attrs['string'] ?? attrs['name'],
        type: btnType,
        invisible: attrs['invisible'],
        groups: attrs['groups'],
      });
    }
  } catch (e) {
    viewButtons = [{ error: String(e) }];
  }

  return {
    ...access,
    server_actions: serverActions.map(a => ({ id: a.id, name: a.name, view_types: a.binding_view_types })),
    reports: reports.map(r => ({ id: r.id, name: r.name, view_types: r.binding_view_types, report_type: r.report_type })),
    view_buttons: viewButtons,
  };
}

// ─── Register ────────────────────────────────────────────────────────────────

export function register(server: McpServer, client: OdooClient, cache: Cache): void {

  server.registerTool(
    'get_models',
    {
      description:
        'List primary models the user can navigate to via menus (get_models()), ' +
        'or list relational models reachable from a base model via its form-view fields ' +
        '(get_models({ base: "sale.order" })).',
      inputSchema: { base: z.string().optional() },
    },
    async ({ base }) => {
      try {
        return ok(base ? await relatedModels(client, cache, base) : await primaryModels(client, cache));
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'get_model_actions',
    {
      description:
        'Return all actions available on a model: server actions (Action menu), ' +
        'report actions (Print menu), and form-view buttons (type=object/action). ' +
        'Also returns CRUD access flags for the current user.',
      inputSchema: {
        model: z.string(),
        action_id: z.number().int().optional(),
      },
    },
    async ({ model, action_id }) => {
      const key = `model_actions:${model}:${action_id ?? ''}`;
      const cached = cache.get(key);
      if (cached) return ok(cached);
      try {
        const result = await collectModelActions(client, model);
        cache.set(key, result);
        return ok(result);
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'get_model_interface',
    {
      description:
        'Single-call planning helper: returns form-view field metadata AND all model actions ' +
        '(server actions, reports, view buttons) AND CRUD access flags. ' +
        'Use this before creating or editing records to understand the full model interface ' +
        'without separate get_fields + get_model_actions round trips.',
      inputSchema: { model: z.string() },
    },
    async ({ model }) => {
      const key = `model_interface:${model}`;
      const cached = cache.get(key);
      if (cached) return ok(cached);
      try {
        const [fields, actions] = await Promise.all([
          client.getFormFields(model),
          collectModelActions(client, model),
        ]);
        const result = { fields, ...actions };
        cache.set(key, result);
        return ok(result);
      } catch (e) { return ok({ error: String(e) }); }
    },
  );
}
